### ISSUE TITLE
Weird behaviour with some full screen ncurses programs in "integrated tmux mode"

### ISSUE THREAD

**Issue by VorpalBlade** [romkatv/zsh4humans#155]
When using certain full screen ncurses program, such as nano, vim, emacs and aptitude outside of zsh4human, the program disables the scrollback in the terminal and also captures the scroll wheel to handle it itself.

However when running under zsh4humans all of these *except aptitude* stop this behaviour and the scrollback in the terminal keeps being enabled and scrolling is handled by the terminal emulator (konsole in my case).

* I checked that this behaviour is reproducible both in the Alpine and Ubuntu docker containers (except aptitude, for obvious reasons I only checked that under Ubuntu).
* I have checked that overriding TERM to be xterm did *not* change the behaviour. So whatever the cause is, it is something else.

I hope that is something that can be fixed or worked around, as I'm quite used to certain behaviour to the point of it being muscle memory basically.

**Comment by romkatv** [comment]
I've reproduced this with GNOME terminal and nano. No ETA for a fix yet.

**Comment by VorpalBlade** [comment]
Another program that is *unaffected* is htop. Maybe that will help figuring out what on earth could cause this. It seems quite inconsistent.

**Comment by romkatv** [comment]
Thanks. This can indeed be useful.

**Comment by romkatv** [comment]
Did a bit of digging. Mouse always works in applications that want mouse: `htop`, `less --mouse`, `nano --mouse`, etc. By "always" I mean that it works in all terminals with and without integrated tmux. Applications that enable mouse not only handle mouse wheel but clicks, too. E.g., with `nano --mouse` clicking moves the cursor to where you've clicked. Selecting text with mouse requires holding <kbd>Shift</kbd>.

Applications that don't request mouse (e.g., `nano` without `--mouse`) don't handle mouse. If you try running them in xterm or Windows Terminal, you'll see that the mouse wheel scrolls the terminal window and allows you to see the scrollback (this would actually be useful if the content above the fold wasn't missing the bottom lines). However, some terminals (VTE, Konsole, term.js, and maybe a few others), implement heuristics that, when triggered, translate mouse wheel to N <kbd>Up</kbd>/<kbd>Down</kbd> keys where N is usually 3 or 5. This doesn't necessarily match the behavior of the mouse wheel when the app handles it explicitly but it's usually close. A notable exception is `tmux` -- the heuristic does kick in for `tmux` if `tmux` doesn't explicitly enable mouse, but sending N <kbd>Up</kbd>/<kbd>Down</kbd> keys to `tmux` when it's displaying zsh prompt is not something that anyone would want. Note that these terminals doesn't translate mouse clicks in its heuristic mode. For some reason integrated tmux breaks the heuristic. I don't know why yet. I'd like to fix this but no ETA.

As a temporary solution I suggest defining aliases that enable mouse explicitly in applications that you care about. This can even bring advantages such as cursor positioning with a mouse click in nano. It'll also enable mouse wheel scrolling in `less -X` (terminal heuristics don't kick in when this option is used). I personally really `-X` and it's one of the default flags in `LESS` in zsh4humans.

**Comment by romkatv** [comment]
> For some reason integrated tmux breaks the heuristic. I don't know why yet. I'd like to fix this but no ETA.

OK, figured it out. The heuristic kicks in when the TTY is in alternate mode and the client has no interest in *any* mouse events. When you run `nano` under integrated tmux, the first condition is satisfied but the second isn't. Integrated tmux subscribes to focus events and those count as mouse events. Subscribing to focus events doesn't make any sense to me but unfortunately I cannot find out from history why I [enabled them](https://github.com/romkatv/zsh4humans/commit/73cb7c8e50373bf04cf8dfb5120c6899e0011291) in the first place. I've disabled them now. Run `z4h update` to pick up the fix and open a new terminal. Let me know if this fixes the issue for you.

**Comment by jared-w** [comment]
> unfortunately I cannot find out from history why I [enabled them](https://github.com/romkatv/zsh4humans/commit/73cb7c8e50373bf04cf8dfb5120c6899e0011291) in the first place

Does `FocusGained` and `FocusLost` autocmd events work in vim/neovim with the integrated tmux setup? [ref](https://github.com/neovim/neovim/issues/9486). If not, that could be why it was originally enabled?

**Comment by romkatv** [comment]
> > unfortunately I cannot find out from history why I [enabled them](https://github.com/romkatv/zsh4humans/commit/73cb7c8e50373bf04cf8dfb5120c6899e0011291) in the first place
> 
> Does `FocusGained` and `FocusLost` autocmd events work in vim/neovim with the integrated tmux setup?

What does this mean? How can you test it?

**Comment by romkatv** [comment]
Closing as fixed.

**Comment by jared-w** [comment]
> > > unfortunately I cannot find out from history why I [enabled them](https://github.com/romkatv/zsh4humans/commit/73cb7c8e50373bf04cf8dfb5120c6899e0011291) in the first place
> > 
> > 
> > Does `FocusGained` and `FocusLost` autocmd events work in vim/neovim with the integrated tmux setup?
> 
> What does this mean? How can you test it?

Vim's autocmd events are analogous to zsh's hooks where you can add a function to a hook and have it run. The `FocusGained` and `FocusLost` events run when the terminal receives/loses focus (eg by focusing a browser window instead). Because tmux is itself "a terminal" (well... the reality is more complicated, as you know), your terminal sending a focus lost/gained event doesn't mean vim will get it if tmux swallows the event and doesn't forward it through to vim.

**Comment by romkatv** [comment]
How can you test whether FocusGained and FocusLost autocmd events work in vim/neovim with the integrated tmux setup?

**Comment by jared-w** [comment]
Sorry, I realized I had forgotten to actually answer the second half. I edited my first comment accordingly.

tl;dr they don't seem to work without tmux subscribing to focus events. But doing so appears to break mouse support heuristics, so I think it's a "pick your poison" scenario here.

**Comment by romkatv** [comment]
Thanks for the instructions how to test it. I agree with your conclusions. If anyone cares about this, I can add an option to enable status events. Ideally, VTE should enable its scrolling heuristic when the foreground process subscribes to status events. It's unclear why it doesn't.
